File: module-app/src/main/java/maple/expectation/service/v2/like/realtime/impl/RedisLikeEventSubscriber.java
package maple.expectation.service.v2.like.realtime.impl;

import io.micrometer.core.instrument.MeterRegistry;
import jakarta.annotation.PostConstruct;
import jakarta.annotation.PreDestroy;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import maple.expectation.infrastructure.cache.TieredCacheManager;
import maple.expectation.infrastructure.executor.LogicExecutor;
import maple.expectation.infrastructure.executor.TaskContext;
import maple.expectation.infrastructure.queue.RedisKey;
import maple.expectation.service.v2.like.realtime.LikeEventSubscriber;
import maple.expectation.service.v2.like.realtime.dto.LikeEvent;
import org.redisson.api.RTopic;
import org.redisson.api.RedissonClient;
import org.redisson.api.listener.MessageListener;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.Cache;

/**
 * Redis RTopic 기반 좋아요 이벤트 구독자
 *
 * <h3>Issue #278: Scale-out 환경 실시간 좋아요 동기화</h3>
 *
 * <p>다른 인스턴스에서 발행한 이벤트를 수신하여 L1(Caffeine) 캐시 무효화
 *
 * <h3>캐시 무효화 전략 (5-Agent Council 합의)</h3>
 *
 * <ul>
 *   <li>Pub/Sub 수신 → L1 즉시 evict
 *   <li>TTL(5분)은 Fallback용 (Pub/Sub 유실 시)
 *   <li>자기 자신이 발행한 이벤트는 무시 (Self-skip)
 * </ul>
 *
 * <h3>CLAUDE.md Section 12: LogicExecutor 패턴</h3>
 *
 * <p>모든 캐시 작업은 executeVoid로 예외 처리
 */
@Slf4j
@RequiredArgsConstructor
public class RedisLikeEventSubscriber implements LikeEventSubscriber {

  private final RedissonClient redissonClient;
  private final TieredCacheManager cacheManager;
  private final LogicExecutor executor;
  private final MeterRegistry meterRegistry;

  @Value("${app.instance-id:${HOSTNAME:unknown}}")
  private String instanceId;

  /** 구독 해제용 리스너 ID */
  private volatile Integer listenerId;

  /** RTopic 인스턴스 (구독 해제용) */
  private volatile RTopic topic;

  /** 이벤트 구독 시작 (애플리케이션 시작 시) */
  @Override
  @PostConstruct
  public void subscribe() {
    TaskContext context = TaskContext.of("LikePubSub", "Subscribe", instanceId);

    executor.executeVoidJava(
        () -> {
          topic = redissonClient.getTopic(RedisKey.LIKE_EVENTS_TOPIC.getKey());

          // MessageListener 등록
          listenerId = topic.addListener(LikeEvent.class, createMessageListener());

          log.info(
              "[LikeEventSubscriber] Subscribed to topic: {}, instanceId={}",
              RedisKey.LIKE_EVENTS_TOPIC.getKey(),
              instanceId);
        },
        context);
  }

  /** 메시지 리스너 생성 (CLAUDE.md Section 15: 람다 3줄 이내) */
  private MessageListener<LikeEvent> createMessageListener() {
    return (channel, event) -> onEvent(event);
  }

  /**
   * 이벤트 수신 및 처리
   *
   * <p>P0-3 대응: Self-skip으로 불필요한 캐시 무효화 방지
   */
  @Override
  public void onEvent(LikeEvent event) {
    // Self-skip: 자기가 발행한 이벤트는 무시
    if (instanceId.equals(event.sourceInstanceId())) {
      log.trace("[LikeEventSubscriber] Self-skip: userIgn={}", event.userIgn());
      return;
    }

    TaskContext context = TaskContext.of("LikePubSub", "OnEvent", event.userIgn());

    executor.executeVoidJava(
        () -> {
          evictL1Cache(event.userIgn());
          recordEventReceived();
          log.debug(
              "[LikeEventSubscriber] L1 cache evicted: userIgn={}, source={}, delta={}",
              event.userIgn(),
              event.sourceInstanceId(),
              event.newDelta());
        },
        context);
  }

  /**
   * L1 캐시 무효화 (character 캐시)
   *
   * <p>TieredCacheManager.getL1CacheDirect()로 L1만 직접 evict
   *
   * <p>L2(Redis)는 모든 인스턴스가 공유하므로 evict 불필요
   */
  private void evictL1Cache(String userIgn) {
    // character 캐시의 L1만 evict (GameCharacter 엔티티)
    Cache characterCache = cacheManager.getL1CacheDirect("character");
    if (characterCache != null) {
      characterCache.evict(userIgn);
    }

    // characterBasic 캐시도 evict (기본 정보)
    Cache basicCache = cacheManager.getL1CacheDirect("characterBasic");
    if (basicCache != null) {
      basicCache.evict(userIgn);
    }
  }

  /** 구독 해제 (애플리케이션 종료 시) */
  @Override
  @PreDestroy
  public void unsubscribe() {
    TaskContext context = TaskContext.of("LikePubSub", "Unsubscribe", instanceId);

    executor.executeVoidJava(
        () -> {
          if (topic != null && listenerId != null) {
            topic.removeListener(listenerId);
            log.info("[LikeEventSubscriber] Unsubscribed from topic: instanceId={}", instanceId);
          }
        },
        context);
  }

  // ==================== Metrics ====================

  private void recordEventReceived() {
    meterRegistry.counter("like.event.received").increment();
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/queue/like/realtime/RedisLikeEventSubscriber.kt
package maple.expectation.infrastructure.queue.like.realtime

import io.micrometer.core.instrument.MeterRegistry
import jakarta.annotation.PostConstruct
import jakarta.annotation.PreDestroy
import maple.expectation.core.dto.like.LikeEvent
import maple.expectation.core.port.out.LikeEventSubscriber
import maple.expectation.infrastructure.cache.TieredCacheManager
import maple.expectation.infrastructure.executor.LogicExecutor
import maple.expectation.infrastructure.executor.TaskContext
import maple.expectation.infrastructure.queue.RedisKey
import org.redisson.api.RedissonClient
import org.redisson.api.listener.MessageListener
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.Cache

/**
 * RTopic 기반 좋아요 이벤트 구독자 (at-most-once)
 *
 * Issue #278: Scale-out 환경 실시간 좋아요 동기화
 *
 * 다른 인스턴스에서 발행한 이벤트를 수신하여 L1(Caffeine) 캐시 무효화
 *
 * 캐시 무효화 전략 (5-Agent Council 합의):
 * - Pub/Sub 수신 → L1 즉시 evict
 * - TTL(5분)은 Fallback용 (Pub/Sub 유실 시)
 * - 자기 자신이 발행한 이벤트는 무시 (Self-skip)
 */
class RedisLikeEventSubscriber(
    private val redissonClient: RedissonClient,
    private val cacheManager: TieredCacheManager,
    private val executor: LogicExecutor,
    private val meterRegistry: MeterRegistry,
    @Value("\${app.instance-id:\${HOSTNAME:unknown}}") private val instanceId: String
) : LikeEventSubscriber {

    private var listenerId: Int = -1

    @PostConstruct
    override fun subscribe() {
        val context = TaskContext.of("LikePubSub", "Subscribe", instanceId)

        executor.executeVoid({
            val topic = redissonClient.getTopic(RedisKey.LIKE_EVENTS_TOPIC.key)
            listenerId = topic.addListener(LikeEvent::class.java) { _, msg ->
                onEvent(msg)
            }
        }, context)
    }

    @PreDestroy
    override fun unsubscribe() {
        if (listenerId >= 0) {
            val context = TaskContext.of("LikePubSub", "Unsubscribe", instanceId)
            executor.executeVoid({
                val topic = redissonClient.getTopic(RedisKey.LIKE_EVENTS_TOPIC.key)
                topic.removeListener(listenerId)
            }, context)
        }
    }

    override fun onEvent(event: LikeEvent) {
        // Self-skip: 자기가 발행한 이벤트는 무시
        if (instanceId == event.sourceInstanceId) {
            return
        }

        val context = TaskContext.of("LikePubSub", "OnEvent", event.userIgn)

        executor.executeVoid({
            evictL1Cache(event.userIgn)
            recordEventReceived()
        }, context)
    }

    /**
     * L1 캐시 무효화 (character 캐시)
     *
     * TieredCacheManager.getL1CacheDirect()로 L1만 직접 evict
     * L2(Redis)는 모든 인스턴스가 공유하므로 evict 불필요
     */
    private fun evictL1Cache(userIgn: String) {
        // character 캐시의 L1만 evict (GameCharacter 엔티티)
        val characterCache: Cache? = cacheManager.getL1CacheDirect("character")
        characterCache?.evict(userIgn)

        // characterBasic 캐시도 evict (기본 정보)
        val basicCache: Cache? = cacheManager.getL1CacheDirect("characterBasic")
        basicCache?.evict(userIgn)

        // characterView 캐시도 evict
        val viewCache: Cache? = cacheManager.getL1CacheDirect("characterView")
        viewCache?.evict(userIgn)
    }

    // ==================== Metrics ====================

    private fun recordEventReceived() {
        meterRegistry.counter("like.event.received").increment()
    }
}
